### PR TITLE
Add sequence_conv_op  and sequence_projection functor

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/4899
fix https://github.com/PaddlePaddle/Paddle/issues/5045

### CODE REVIEW

**Comment by chengduoZH** [comment]
Because seq_project is only used in seq_conv, seq_project should be written in functor form.

**Comment by dzhwinter** [comment]
I think we can merge it first and review the code the same time. @chengduoZH Please continue to polish the code based on the comments.

**Comment by chengduoZH** [comment]
@dzhwinter Ok!

File: paddle/operators/sequence_conv_op.cc [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    int context_start = ctx->Attrs().Get<int>("context_start");
+
+    auto in_dims = ctx->GetInputDim("X");
+    auto filter_dims = ctx->GetInputDim("Filter");
+    PADDLE_ENFORCE(in_dims.size() == 2 && filter_dims.size() == 2,
+                   "Input(X, Filter) should be 2-D tensor.");
+    PADDLE_ENFORCE(
+        filter_dims[0] == context_length && filter_dims[1] == in_dims[1],
+        "Filter's shape should be (context_length x "
+        "number_of_input_features).");
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: The filter shape is not right. 

假如：context_length = 3, 输入hidden size = D， 输出的hidden size = H
Filter: [3D, H]

File: paddle/operators/sequence_conv_op.cc [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      PADDLE_ENFORCE(padding_dim.size() == 2,
+                     "Input(PaddingData) should be 2-D tensor.");
+      PADDLE_ENFORCE(
+          padding_dim[0] == total_pad && padding_dim[1] == input_width,
+          "Input(PaddingData)'s shape is not consistent with 'context_start' "
+          "and 'context_length'.");
+    }
+
+    in_dims[1] = 1;
+    ctx->SetOutputDim("Out", in_dims);
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: The output shape is not right.

依据上面假设输出dims[1] = H。

File: paddle/operators/sequence_conv_op.cc [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+};
+
+class SequenceConvOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  SequenceConvOpMaker(framework::OpProto* proto,
+                      framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X",
+             "(A float LoDTensor) the input of SequenceConvOp, a vector of "
+             "2-D matrix of size (minibatch, number_of_input_features).");
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: the first dimension is not minibatch size, it is the total time steps. 可以参考lstm_op的文档。

File: paddle/operators/sequence_conv_op.cc [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    AddInput("PaddingData",
+             "(Tensor) the input of SequenceConvOp, a vector of "
+             "2-D matrix of size (up_pad + down_pad, "
+             "number_of_input_features). ")
+        .AsDispensable();
+    AddInput("Filter",
+             "(Tensor) the input of SequenceConvOp, a vector of "
+             "2-D matrix of size (context_length x number_of_input_features).");
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: The shape is not right, same as above.

File: paddle/operators/sequence_conv_op.cc [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    AddOutput("Out",
+              "(A float LoDTensor) the output of SequenceConvOp, a vector "
+              "of 2-D matrix of size (minibatch, 1).");
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: The output width is not right.

File: paddle/operators/sequence_conv_op.cc [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+    AddAttr<bool>("padding_trainable",
+                  "(bool, default false) the padding data of SequenceConvOp "
+                  "is trainable or not.")
+        .SetDefault(false);
+    AddAttr<int>("context_length",
+                 "(int, default 3) the context_length of SequenceConvOp.")
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: What is context_length? Not explain clearly.

Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: What is context_length? Not explain clear.

File: paddle/operators/sequence_conv_op.cc [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        .SetDefault(3)
+        .GreaterThan(0);
+    AddAttr<int>("context_start",
+                 "(int, default 0) the context_start of SequenceConvOp.")
+        .SetDefault(0);
+    AddAttr<int>("context_stride",
+                 "(int, default 1) the context_stride of SequenceConvOp. "
+                 "Currently, sequence_project_op only support "
+                 "context_stride=1.")
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: The explanation of  `context_start` and `context_stride` is also not clearly.

File: paddle/operators/sequence_conv_op.cc (hunk 2 above)
Comment (qingqing01) [reply paddle/operators/sequence_conv_op.cc]: Also should set LoD for output.

File: paddle/operators/sequence_conv_op.h [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+template <typename Place, typename T>
+class SequenceConvKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* in = context.Input<LoDTensor>("X");
+    auto* out = context.Output<LoDTensor>("Out");
+    auto filter = *context.Input<Tensor>("Filter");
+
+    out->mutable_data<T>(context.GetPlace());
+    //  out->set_lod(in->lod());
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.h]: CompileTimeInferShape等代码加进来之后，导致在InferShape里的`ShareLoD()`函数没有使能，当前为了输出也包含LoD，`Compute`里也调用下`ShareLoD()`吧，可以参考lstm_op。

File: paddle/operators/sequence_conv_op.h [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+    // use col_shape in the im2col calculation
+    framework::DDim col_shape = {in->dims()[0],
+                                 sequence_width * context_length};
+    Tensor col;
+    col.mutable_data<T>(col_shape, context.GetPlace());
+    // Because if padding_trainable is false, padding data should be zeros.
+    auto temp = framework::EigenVector<T>::Flatten(col);
+    temp.device(context.GetEigenDevice<Place>()) =
+        temp.constant(static_cast<T>(0));
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.h]: 调用`math:: SetConstant`置零： https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/math/math_function.h#L97

File: paddle/operators/sequence_conv_op.h [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    LoDTensor* input = const_cast<LoDTensor*>(in);
+    Tensor* pad_data = const_cast<Tensor*>(padding_data);
+
+    seq_project_functor(context.device_context(), *input, *pad_data, col,
+                        padding_trainable, context_start, context_length,
+                        context_stride, up_pad, down_pad, false, false, false);
+
+    filter.Resize(framework::make_ddim({context_length * sequence_width, 1}));
+    math::matmul<Place, T>(context.device_context(), col, false, filter, false,
+                           T(1.0), out, T(0.0));
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.h]: `T(1.0)` -> `static_cast<T>(1.0)`

File: paddle/operators/sequence_conv_op.h [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    framework::DDim col_shape = {in->dims()[0],
+                                 sequence_width * context_length};
+    Tensor col;
+
+    if (in_g || filter_g || (padding_trainable && padding_data_g)) {
+      col.mutable_data<T>(col_shape, context.GetPlace());
+      // Because if padding_trainable is false, padding data should be zeros.
+      auto temp = framework::EigenVector<T>::Flatten(col);
+      temp.device(context.GetEigenDevice<Place>()) =
+          temp.constant(static_cast<T>(0));
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.h]: 调用`math:: SetConstant`置零： https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/math/math_function.h#L97

File: paddle/operators/sequence_conv_op.h [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    if (filter_g) {
+      filter_g->mutable_data<T>(context.GetPlace());
+
+      math::SetConstant<Place, T> functor;
+      functor(context.device_context(), filter_g, 0);
+
+      Tensor filter_grad_ = *filter_g;
+      LoDTensor out_grad_ = *out_g;
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.h]: `out_grad_` -> `out_grad`

File: python/paddle/v2/framework/tests/test_seq_conv.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+
+                in_sub = x[in_begin:in_end, :]
+                out[out_begin:out_end, j * self.input_size[1]:(j + 1) *
+                    self.input_size[1]] += in_sub
+
+        filter_dim = filter.shape
+        output_dim = self.outputs['Out'].shape
+        filter.shape = filter_dim[0] * filter_dim[1]
+        self.outputs['Out'].shape = (output_dim[0], )
+        np.dot(out, filter, out=self.outputs['Out'])
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_seq_conv.py]: Python单测forward实现，觉得避免和C++ Code一致，避免采用先展开后矩阵乘的形式，可以是Conv原本实现形式。

File: paddle/operators/sequence_conv_op.cc (hunk 6 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.cc]: Done

File: paddle/operators/sequence_conv_op.cc (hunk 7 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.cc]: Done

File: paddle/operators/sequence_conv_op.h (hunk 9 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.h]: Done

File: paddle/operators/sequence_conv_op.h (hunk 11 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.h]: Done

File: python/paddle/v2/framework/tests/test_seq_conv.py (hunk 13 above)
Comment (chengduoZH) [reply python/paddle/v2/framework/tests/test_seq_conv.py]: Python单测是根据之前paddle改写过来的，context_project_functor是先经过im2col再通过矩阵乘得到的，这两种方式并不太一样

File: paddle/operators/math/context_project.h [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+template <typename Place, typename T>
+class ContextProjectFunctor {
+ public:
+  void operator()(const platform::DeviceContext& context,
+                  framework::LoDTensor& in, framework::Tensor& padding_data,
+                  framework::Tensor& col, bool padding_trainable,
+                  int context_start, int context_length, int context_stride,
+                  int up_pad, int down_pad, bool gradient, bool input_grad,
+                  bool pad_grad) {
```
Comment (qingqing01) [inline paddle/operators/math/context_project.h]: 觉得将projection和un-projection的过程混合在一起，代码逻辑不够清晰。

File: paddle/operators/math/context_project.h [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+/*
+ * \brief Context projection concatenate features in adjacent time steps in
+ * a sequence. The i-th row of the output is the concatenation of
+ * context_length rows of the input. The context_length rows are the
+ * consecutive rows from the i+shift_start row.
+
+ * \param in            Input data.
+ * \param Shape         The shape of Input data,
+ *                      [minibatch, number_of_input_features].
+ * \param type          A float LoDTensor.
```
Comment (qingqing01) [inline paddle/operators/math/context_project.h]: Remove the type, there is no meaning here.

The argument type in the following function is clear.

File: paddle/operators/math/context_project.h [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+/*
+ * \brief Context projection concatenate features in adjacent time steps in
+ * a sequence. The i-th row of the output is the concatenation of
+ * context_length rows of the input. The context_length rows are the
+ * consecutive rows from the i+shift_start row.
+
+ * \param in            Input data.
+ * \param Shape         The shape of Input data,
+ *                      [minibatch, number_of_input_features].
```
Comment (qingqing01) [inline paddle/operators/math/context_project.h]: number_of_input_features -> input_hidden_size

File: paddle/operators/sequence_conv_op.cc [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+             "This is a tensor with shape (N, D), where N is the "
+             "context_length, D is the output feature size.");
+    AddOutput(
+        "Out",
+        "(LoDTensor) the output(Out) is a LodTensor, which support "
+        "variable-time length output sequence. The underlying tensor in "
+        "this LoDTensor is a matrix with shape (T, D), where, T is the "
+        "total time steps in this mini-batch, D is the output feature size.");
+
+    AddAttr<bool>("padding_trainable",
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: `paddingTrainable`, please to see our name convention.

File: paddle/operators/sequence_conv_op.cc [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+                  "(bool, default false) the padding data of SequenceConvOp "
+                  "is trainable or not.")
+        .SetDefault(false);
+    AddAttr<int>("context_length",
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: `contextLength `

File: paddle/operators/sequence_conv_op.cc [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+                 "(int, default 3) the context_length of SequenceConvOp is the "
+                 "height of the convolution kernel.")
+        .SetDefault(3)
+        .GreaterThan(0);
+    AddAttr<int>("context_start",
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: `contextStart `

File: paddle/operators/sequence_conv_op.cc [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+                 "(int, default 0) the context_start of SequenceConvOp "
+                 "represents the beginning of the convolution of the number of "
+                 "rows of sequence, which can be negative.")
+        .SetDefault(0);
+    AddAttr<int>("context_stride",
```
Comment (qingqing01) [inline paddle/operators/sequence_conv_op.cc]: `contextStride `

File: python/paddle/v2/framework/tests/test_seq_conv.py [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+        self.context_length = 3
+        self.padding_trainable = True
+        self.context_stride = 1
+
+        self.input_size = [self.input_row, 23]
+        idx = range(self.input_size[0])
+        del idx[0]
+        self.lod = [[0] + np.sort(random.sample(idx, 8)).tolist() +
+                    [self.input_size[0]]]
+        self.output_represention = 8  # output feature size
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_seq_conv.py]: Need unit testing for the case `self.context_stride > 1`

File: paddle/operators/math/context_project.h (hunk 16 above)
Comment (chengduoZH) [reply paddle/operators/math/context_project.h]: Done. https://github.com/PaddlePaddle/Paddle/pull/5130

File: paddle/operators/math/context_project.h (hunk 15 above)
Comment (chengduoZH) [reply paddle/operators/math/context_project.h]: Done. https://github.com/PaddlePaddle/Paddle/pull/5130

File: paddle/operators/sequence_conv_op.cc (hunk 17 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.cc]: Done. https://github.com/PaddlePaddle/Paddle/pull/5130

File: paddle/operators/sequence_conv_op.cc (hunk 18 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.cc]: Done. https://github.com/PaddlePaddle/Paddle/pull/5130

File: paddle/operators/sequence_conv_op.cc (hunk 19 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.cc]: Done. https://github.com/PaddlePaddle/Paddle/pull/5130

File: paddle/operators/sequence_conv_op.cc (hunk 20 above)
Comment (chengduoZH) [reply paddle/operators/sequence_conv_op.cc]: Done. https://github.com/PaddlePaddle/Paddle/pull/5130

File: python/paddle/v2/framework/tests/test_seq_conv.py (hunk 21 above)
Comment (chengduoZH) [reply python/paddle/v2/framework/tests/test_seq_conv.py]: Currently, seq_conv_op only supports self.context_stride = 1.

File: paddle/operators/math/context_project.h (hunk 14 above)
Comment (chengduoZH) [reply paddle/operators/math/context_project.h]: 分开写也是可以的，不过显得代码有点冗余，我再想想办法

**Pull request merged**